Support for non-keyed .attrs factory function

oliverlaz opened the pull request:
This PR enhances the ``.attrs`` function support for a top level factory function which receives the component props as an argument.

Example:
```javascript
const Input = styled.input.attrs(props => ({
  placeholder: props.placeholder
}))`
  color: red;
`
```

This change is very handy when integrating a styled-component within [``redux-form``](https://redux-form.com) [``Field``](https://redux-form.com/7.0.4/docs/api/field.md/) element where this code:
```javascript
const Input = styled.input.attrs({
  onChange: props => props.input.onChange,
  onBlur: props => props.input.onBlur,
  onFocus: props => props.input.onFocus,
  ...
})`
  color: red;
`
```

can be easily changed to:
```javascript
const Input = styled.input.attrs(props => ({
   ...props.input // ensure all redux-form props and handlers are passed to the wrapped input component
})`
  color: red;
`
```

philpl commented:
cc @geelen; this is something I wanted to add some time ago 😆

philpl commented on src/models/StyledComponent.js:
```diff
@@ -64,14 +64,22 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
+      return Object.keys(attrs).reduce((acc, key) => {
         const attr = attrs[key]
         // eslint-disable-next-line no-param-reassign
         acc[key] = typeof attr === 'function' ? attr(context) : attr
```
this should probably be inactive if the alternative syntax is used. Otherwise you can do: `p => ({ a: p => T })` which really has no point

geelen commented:
Yeah happy with this change and agree it should be either/or, top level or
per attr functions, but not both
On Tue, 19 Sep 2017 at 10:15 pm Phil Plückthun <notifications@github.com>
wrote:

> *@philpl* commented on this pull request.
> ------------------------------
>
> In src/models/StyledComponent.js
> <https://github.com/styled-components/styled-components/pull/1171#discussion_r139673341>
> :
>
> > @@ -64,14 +64,22 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
>          return context
>        }
>
> -      this.attrs = Object.keys(attrs).reduce((acc, key) => {
> +      this.attrs = this.resolveComponentAttrs(attrs, context)
> +
> +      return { ...context, ...this.attrs }
> +    }
> +
> +    resolveComponentAttrs(attrs: any, context: any) {
> +      if (typeof attrs === 'function') {
> +        return attrs(context)
> +      }
> +
> +      return Object.keys(attrs).reduce((acc, key) => {
>          const attr = attrs[key]
>          // eslint-disable-next-line no-param-reassign
>          acc[key] = typeof attr === 'function' ? attr(context) : attr
>
> this should probably be inactive if the alternative syntax is used.
> Otherwise you can do: p => ({ a: p => T }) which really has no point
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/styled-components/styled-components/pull/1171#pullrequestreview-63637820>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABa4Gdzmoh_BiU4Jml1xAvf7GQc-hnwks5sj7BagaJpZM4PbnvO>
> .
>

philpl commented on src/constructors/constructWithOptions.js:
```diff
@@ -18,9 +18,13 @@ export default (css: Function) => {
-        attrs: { ...(options.attrs || {}), ...attrs } })
+        attrs: typeof attrs === 'function'
+          ? (context) => ({ ...(options.attrs || {}), ...attrs(context) })
+          : { ...(options.attrs || {}), ...attrs } })
```
This doesn't cover `options.attrs` being mixed. It's making the assumption that the previous one has the same type, but `options.attrs` might either be a function producing an object or an object as well.

philpl commented on src/models/StyledComponent.js:
```diff
@@ -64,14 +64,22 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
+      return { ...context, ...this.attrs }
+    }
+
+    resolveComponentAttrs(attrs: any, context: any) {
```
`attrs: any` is defeating flow types here

philpl commented on src/models/StyledComponent.js:
```diff
@@ -64,14 +64,22 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
+        return attrs(context)
+      }
+
+      return Object.keys(attrs).reduce((acc, key) => {
```
I'm wondering... We could abstract this part of the `attrs` resolution to `constructWithOptions` now. This would help fixing the above described edge case and would result in always being able to run `attrs(context)` here, which helps v8 in optimising it.

mxstbr commented:
> it should be either/or, top level or per attr functions, but not both

Let's support both in the next major release (v3) but log a warning when used per-attr that warns folks to upgrade and then in the major release after that (v4) we deprecate and remove the per-attr version?

oliverlaz commented:
hey @philpl, thanks for your findings! I've changed the code according to your proposals and added few more tests.
Can you have another look at this?

@mxstbr, I agree with you, the support for per-attr props shouldn't be immediately dropped. Having a migration path is always a good approach. 
Please let me know if deprecation warnings needs to be added. I will be happy to help.

nelix commented:
I was just writing a babel macro for this...  Really dont know why i didnt consider making a PR.. lol

philpl commented on src/constructors/constructWithOptions.js:
```diff
@@ -10,6 +10,21 @@ export default (css: Function) => {
+    // Currently it supports both ways for defining component attrs:
+    // - plain object -> styled.input.attrs({ example: 'value' })
+    // - factory function -> styled.input.attrs(props => ({ example: props.value })
+    const createAttrsResolver = attrs => context => {
```
Can you turn this into a pure function please? options.attrs can be passed separately. Also you can simplify the logic and prevent future bugs by not adding resolveAttrs, but by just using the attrs option. In that case you’d wrap an object in a function that just takes the context, and then only have the function-only combination logic here. That would also remove one branch

mxstbr commented:
Ping @oliverlaz, mind addressing that one last comment above so we can ship this? I'd love to have this so we can start logging warnings if used the other way in v3.

oliverlaz commented:
Hi @mxstbr, I will try to find time to finish this until end of this week. I hope this is fine for you.

mxstbr commented:
Perfect, thanks so much @oliverlaz and sorry to let this sit for so long!

jesstech commented:
Any progress on this?

jesstech commented:
@probablyup Unfortunately I need to maintain a `styled()` instance, so it would seem that wrapping with a HOC is infeasible.

kitten commented:
@oliverlaz Hiya, sorry for not keeping this on track. We'd love to get this into the next minor version with all of our other changes. Are you still available and happy to work on this? Otherwise I'd be happy to rebase and finish it up for you :wink:

oliverlaz commented:
hey @kitten, I'm always happy to help, but right now I hardly find time to tackle this.
I also don't like this feature to hang and I will be happy if you or somebody else pushes it forward :)

kitten commented:
@oliverlaz cool! no worries, I'll try to pick it up if I can find time next week :+1:

probablyup commented:
@oliverlaz still swamped? If you have some time to get this rebased and ready I'd be happy to land it

probablyup commented:
@oliverlaz if you have time, i'd love to get this rebased against the `develop` branch and potentially land it in v4 or v4.1

oliverlaz commented:
Hey @probablyup, I’m on a vacation at the moment, away from a laptop :)
I could find some time to continue working on this in two weeks from now.

If you need this sooner, feel free to hand this over to someone else.

Btw, thx for the great work with the v4 release!

probablyup commented:
Thanks! Yeah that’s no problem, we can plan for it to land for 4.1
On Wed, Sep 12, 2018 at 9:45 AM Oliver Lazoroski <notifications@github.com>
wrote:

> Hey @probablyup <https://github.com/probablyup>, I’m on a vacation at the
> moment, away from a laptop :)
> I could find some time to continue working on this in two weeks from now.
>
> If you need this sooner, feel free to hand this over to someone else.
>
> Btw, thx for the great work with the v4 release!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/styled-components/styled-components/pull/1171#issuecomment-420674971>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAiy1n3u6bS27gEfRSQM0cZpUO5EsN9eks5uaR3tgaJpZM4PbnvO>
> .
>

probablyup commented:
Ping @oliverlaz, any chance you have time to pick this up? Would love to include it in 4.1 if possible.

Fer0x commented:
@probablyup I have started to picking this up, but stuck a little with proper merging. I will keep on this week.

oliverlaz commented:
Hi @probablyup, @Fer0x,

I also run into some issues with rebasing yesterday so I started from scratch. I have something already done and working. I will open a follow-up PR about that in the next few hours.

There is one issue that I still have to solve, related to `ComponentStyle.isStatic` resolution.

Fer0x commented:
@oliverlaz Ok, then we are waiting for your new PR. 

You can look at my version for new `isStatic` https://github.com/styled-components/styled-components/blob/props-in-attrs/src/utils/isStaticRules.js#L20

I passed attrs as array because in some cases you can have mix of function and object attrs (i.e. from constructor and `target.attrs`).

Fer0x commented:
Close because of #2176
